fix: new fork 'withdrawals' and fixes for asset unlock signature verification

knst opened the pull request:
## Issue being fixed or feature implemented
https://github.com/dashpay/dash-issues/issues/83

## What was done?
Introduces new fork "withdrawals" which let Asset Unlock be valid for any active quorum + 1 extra inactive (in opposite of hard-coded 2 of them).


## How Has This Been Tested?
See new test section `test_withdrawals_fork` in functional test feature_asset_locks.py

## Breaking Changes
 - new fork "withdrawals"
 - new logic of validation of Asset Unlock transactions's signature. Even no asset-unlock is mined yet, previous version of clients will refuse blocks which is fine for current system



## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

thephez commented on src/chainparams.cpp:
```diff
@@ -208,15 +209,14 @@ class CMainParams : public CChainParams {
-        consensus.vDeployments[Consensus::DEPLOYMENT_MN_RR].nFalloffCoeff = 5;          // this corresponds to 10 periods
-        consensus.vDeployments[Consensus::DEPLOYMENT_MN_RR].useEHF = true;
+        consensus.vDeployments[Consensus::DEPLOYMENT_WITHDRAWALS].bit = 11;
+        consensus.vDeployments[Consensus::DEPLOYMENT_WITHDRAWALS].nStartTime = 1704067200;   // January 1, 2024
+        consensus.vDeployments[Consensus::DEPLOYMENT_WITHDRAWALS].nTimeout = 1767225600; // January 1, 2026
```
Start time and timeout values for each network should be updated to something more relevant before merging

UdjinM6 replied:
1. `nStartTime` should be somewhere in the (near) future
2. `nStartTime` and `nTimeout` should be 1 year apart on mainnet

PastaPastaPasta commented:
Needs rebase?

knst commented:
> Needs rebase?
depends on #6275 (bury mn_rr) also, which is not merged yet; will rebase after

UdjinM6 commented on src/evo/assetlocktx.cpp:
```diff
@@ -113,13 +114,18 @@ bool CAssetUnlockPayload::VerifySig(const llmq::CQuorumManager& qman, const uint
+    assert(llmq_params_opt.has_value());
+
+    // after deployment WITHDRAWALS activated we check not to quorum, but all active quorums + 1 the latest inactive
+    const int quorums_to_scan = DeploymentActiveAt(*pindexTip, Params().GetConsensus(), Consensus::DEPLOYMENT_WITHDRAWALS) ? (llmq_params_opt->signingActiveQuorumCount + 1) : 2;
```
could use `keepOldConnections` here
```suggestion
    const int quorums_to_scan = DeploymentActiveAt(*pindexTip, Params().GetConsensus(), Consensus::DEPLOYMENT_WITHDRAWALS) ? llmq_params_opt->keepOldConnections : 2;
```

knst replied:
> could use keepOldConnections here

technically, yes

But logically I can't explain myself how  `signingActiveQuorumCount` and `keepOldConnections` are related.

UdjinM6 replied:
```
    // Used for intra-quorum communication. This is the number of quorums for which we should keep old connections.
    // For non-rotated quorums it should be at least one more than the active quorums set.
    // For rotated quorums it should be equal to 2 x active quorums set.
    int keepOldConnections;
```
https://github.com/dashpay/dash/blob/develop/src/llmq/params.h#L102-L105

i.e. active set + 1 quorum that became inactive recently, exactly what you need here

knst replied:
I'd still says, that in some particular cases it is matched with "+1" indeed, but I can't justify use "keep old connections" as amount of quorums which makes signature valid; also "2x active quorums" for rotating is not relevant for this situation.

UdjinM6 commented:
tests failed

knst commented:
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/7bc369813b6d64384faef08b16448515e1619c05..5fd7b07ddc24fe5fbb2bebbf8e38d6b92f1ddff8) the fix-withdrawal-quorums branch from[ 7bc3698 ](https://github.com/dashpay/dash/commit/7bc369813b6d64384faef08b16448515e1619c05)to 5fd7b07

fixed test, activation date and rebased to the newer commit - base commit is pretty old

UdjinM6 reviewed (changes requested):
pls see https://github.com/UdjinM6/dash/commits/pr_6279/

knst commented:
```
int CAssetUnlockPayload::getHeightToExpiry() const
{
    const auto& llmq_params_opt = Params().GetLLMQ(Params().GetConsensus().llmqTypePlatform);
    assert(llmq_params_opt.has_value());
    return requestedHeight + llmq_params_opt->dkgInterval * (llmq_params_opt->signingActiveQuorumCount + 1);
}
```
@UdjinM6 no, accordingly DIP and by design, `getHeightToExpire` is not related to the amount of active quorums at all, it's a mechanism to refuse withdrawal in a short time (48 blocks) if it has not been mined, for example due to limits, or due to small fee; so it can be re-issued without waiting 24+ hours

UdjinM6 reviewed (changes requested):
pls see 0c8b3778d7

UdjinM6 commented on test/functional/feature_asset_locks.py:
```diff
@@ -646,8 +646,11 @@ def test_withdrawal_fork(self, node_wallet, pubkey):
         index = 501
-        while True:
-            self.log.info(f"Generating new Asset Unlock tx, index={index}...")
+        attempts = 0
+        while attempts < 10
+            attempts = attempts + 1
+
+            self.log.info(f"Generating new Asset Unlock tx, index={index}, attempt={attempts}...")
```
Good catch! Could reuse `index` though e.g.
```suggestion
        while index < 511:
            self.log.info(f"Generating new Asset Unlock tx, index={index}...")
```

UdjinM6 reviewed (changes requested):
invalid syntax in `while attempts < 10` (missing`:`), also pls see below

UdjinM6 reviewed (approved):
utACK d69030951389d5d2c160e13024c86ca815a5a5e4

PastaPastaPasta commented on src/evo/assetlocktx.cpp:
```diff
@@ -113,13 +114,21 @@ bool CAssetUnlockPayload::VerifySig(const llmq::CQuorumManager& qman, const uint
+    const auto& llmq_params_opt = Params().GetLLMQ(llmqType);
+    assert(llmq_params_opt.has_value());
+
+    // We check two quorums before DEPLOYMENT_WITHDRAWALS activation
+    // and "all active quorums + 1 the latest inactive" after activation.
+    const int quorums_to_scan = DeploymentActiveAt(*pindexTip, Params().GetConsensus(), Consensus::DEPLOYMENT_WITHDRAWALS)
+                                    ? (llmq_params_opt->signingActiveQuorumCount + 1)
+                                    : 2;
+    const auto quorums = qman.ScanQuorums(llmqType, pindexTip, quorums_to_scan);
 
     if (bool isActive = std::any_of(quorums.begin(), quorums.end(), [&](const auto &q) { return q->qc->quorumHash == quorumHash; }); !isActive) {
-        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-assetunlock-not-active-quorum");
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-assetunlock-too-old-quorum");
```
why change this?

PastaPastaPasta reviewed (approved):
utACK d69030951389d5d2c160e13024c86ca815a5a5e4
